### PR TITLE
feat(ctb): Add new move type to FDG for OR counters

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Adds a potential new move type to the `FaultDisputeGame` that allows for a participant to reveal the preimage of the claimed output root to display to the dispute game that the claimed L2 block number does not match up with the block number that the block header within the output root commits to.

### Semantics

The root output can be challenged with the new special move type iff:
1. The passed `OutputRootProof` hashes to equal the claimed output root.
1. The passed Header RLP hashes to equal the block hash within the `OutputRootProof` above.
1. The claimed block number in the dispute game does not equal the block number that the output root commits to.

If there is a successful challenge with the new move type, that claim itself is inserted as a special case counter. In `resolveClaim`, the contract will always consider the creator of the L2 block challenge the winner of the bond. Notably, this only applies for the root claim subgame.

### CODE REVIEW

**Comment by clabby** [comment]
* **#10439** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/10439?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#10438** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/10438?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @clabby and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by Inphi — COMMENTED** [review]
This is good. Sorry to ask you to do this, but can we avoid touching `claimData` or introducing new claims?

The changes to resolution should be minimal; just need to add a check asserting that the root's subgame cannot be resolved if the L2 block number is challenged.

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol [hunk 1]
```diff
@@ -462,6 +477,51 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
...
+        bytes memory rawBlockNumber = RLPReader.readBytes(headerContents[HEADER_BLOCK_NUMBER_INDEX]);
+
+        // Sanity check the block number string length.
+        if (rawBlockNumber.length > 32) revert InvalidHeaderRLP();
+
+        // Convert the raw, left-aligned block number to a uint256 by aligning it as a big-endian
+        // number in the low-order bytes of a 32-byte word.
+        //
+        // SAFETY: The length of `rawBlockNumber` is checked above to ensure it is at most 32 bytes.
+        uint256 blockNumber;
```
Comment (smartcontracts) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol:518]: Can use the function provided by the RLP library instead

**Review by smartcontracts — COMMENTED** [review]
Looking good, leaving one comment. Personally I would've preferred not to make the RLP changes but they've been made already so whatever. We need to justify any diff that's being made on top of audited code when we go to gov vote, prefer to avoid this sort of stuff if we can.

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol (hunk 1 above)
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol:518]: There isn't one, it looks like

**Review by refcell — APPROVED** [review]
Looks great to me! Nice unit tests as well

**Review by Inphi — APPROVED** [review]
Excellent!

**Pull request merged**